clip/model.py

```diff
@@ -0,0 +1,447 @@
+# Copyright © 2023 Apple Inc.
+
+import json
+from dataclasses import dataclass
+from functools import reduce
+from pathlib import Path
+from typing import Any, List, Optional, Union
+
+import mlx.core as mx
+import mlx.nn as nn
+from config import CLIPConfig, CLIPTextConfig, CLIPVisionConfig
+from huggingface_hub import hf_hub_download
+from mlx.core import linalg as LA
+from mlx.nn.losses import cross_entropy
+from mlx.utils import tree_flatten
+
+
+def quick_gelu(x: mx.array) -> mx.array:
+    """
+    Applies GELU approximation that is fast but somewhat inaccurate. See: https://github.com/hendrycks/GELUs
+    """
+    return x * mx.sigmoid(1.702 * x)
+
+
+def clip_loss(logits_per_text: mx.array, logits_per_image: mx.array) -> mx.array:
+    [N, _] = logits_per_text.shape
+    [M, _] = logits_per_image.shape
+    caption_loss = cross_entropy(logits_per_text, mx.arange(N), reduction="mean")
+    image_loss = cross_entropy(logits_per_image, mx.arange(M), reduction="mean")
+    return (caption_loss + image_loss) / 2.0
+
+
+def normal(shape: List[int], mean=0.0, std=1.0) -> mx.array:
+    return mean + std * mx.random.normal(shape)
+
+
+@dataclass
+class CLIPVisionOutput:
+    pooler_output: mx.array
+    last_hidden_state: mx.array
+
+
+@dataclass
+class CLIPTextOutput:
+    pooler_output: mx.array
+    last_hidden_state: mx.array
+
+
+@dataclass
+class CLIPModelOutput:
+    loss: Optional[mx.array]
+    text_embeds: Optional[mx.array]
+    image_embeds: Optional[mx.array]
+    text_model_output: CLIPTextOutput
+    vision_model_output: CLIPVisionOutput
+
+
+class CLIPEncoderLayer(nn.TransformerEncoderLayer):
+    """The transformer encoder layer from CLIP."""
+
+    def __init__(self, hidden_dim: int, intermediate_dim: int, num_heads: int):
+        super().__init__(
+            dims=hidden_dim,
+            mlp_dims=intermediate_dim,
+            num_heads=num_heads,
+            activation=quick_gelu,
+            norm_first=True,
+        )
+        self.attention.query_proj.bias = mx.zeros(hidden_dim)
+        self.attention.key_proj.bias = mx.zeros(hidden_dim)
+        self.attention.value_proj.bias = mx.zeros(hidden_dim)
+        self.attention.out_proj.bias = mx.zeros(hidden_dim)
+
+    def init_weights(self, factor: float, num_hidden_layers: int):
+        [embed_dim, _] = self.linear2.weight.shape
```

Comment:
Same as above.

Reply:
I will rewrite to match your suggestion.